### ISSUE TITLE
How to make log context work in webflux and spring sleuth - let's discuss alternatives

### ISSUE THREAD

**Issue by ahus1** [ahus1/spring-webflux-with-log-context#1]
Hello @simonbasle, 

thanks for publishing your article about MDC and WebFlux: https://simonbasle.github.io/2018/02/contextual-logging-with-reactor-context-and-mdc/

You end your example with Mono - I wonder if you can stay with a Flux, and how to integrate with Spring Sleuth. 

I tried to use Reactor's Hook with no luck, and ended up with wrapping the Runnables and Consumers for the doXXX methods. 

See [README.adoc](https://github.com/ahus1/spring-webflux-with-log-context/blob/master/README.adoc) and [RestaurantController.java](https://github.com/ahus1/spring-webflux-with-log-context/blob/master/src/main/java/de/ahus1/reactive/context/demo/RestaurantController.java) for the code.

I discussed similar problems previously with @adriancole and @thomasdarimont - maybe they want to join in.

Maybe even @marcingrzejszczak wants to join in - I'd be honored.

Regards,
Alexander aka @ahus1

**Comment by thomasdarimont** [comment]
Another "poor" mans solution would be to use an aspect to apply the wrapping logic from your example.

**Comment by marcingrzejszczak** [comment]
Sleuth already works with reactor and flux

**Comment by simonbasle** [comment]
@ahus1 I use `Mono.just` in my example, but that just so that the initial log statement can be performed using `doOnEach`. There's no real need for that as long as you want to perform log statements that match the `onNext` events in your sequence.

**Comment by madgnome** [comment]
@simonbasle In your solution, you use `doOnEach` which call `subscribe` on the source internally so when you have this type of code in your Controller

```
restaurantService.byPrice(maxPrice))
	           .doOnEach(sig -> logOnNext(sig, r -> LOG.debug("found restaurant {} for ${}", r.getName(), r.getPricePerPerson())))
```

you end up with `restaurantService.byPrice` being subscribed twice (once by spring, once by doOnEach)

**Comment by simonbasle** [comment]
No, it calls subscribe _when it is subscribed to_, like every other operator. Since you return the `doOnEach` to Spring, Spring will subscribe to doOnEach which will in turn subscribe to `byPrice`.

**Comment by madgnome** [comment]
Ok, maybe I haven't read the source of `MonoDoOnEach` correctly then:

```
final class MonoDoOnEach<T> extends MonoOperator<T, T> {

	final Consumer<? super Signal<T>> onSignal;

	MonoDoOnEach(Mono<? extends T> source, Consumer<? super Signal<T>> onSignal) {
		super(source);
		this.onSignal = Objects.requireNonNull(onSignal, "onSignal");
	}

	@Override
	public void subscribe(CoreSubscriber<? super T> actual) {
		//TODO fuseable version?
		//TODO conditional version?
		source.subscribe(new FluxDoOnEach.DoOnEachSubscriber<>(actual, onSignal));
	}
}
```

**Comment by simonbasle** [comment]
@madgnome yes, notice that subscribe call is in the subscribe method. Using `Flux` operators in a fluent style (as they are intended to be used) creates a chain of operator implementations like `MonoDoOnEach` that each reference the previous step in the chain (the `source`).
But nothing happens until you subscribe to the last instance in the chain. At this point, the subscription signal is propagated backwards, creating a chain of subscribers: the `actual` in `subscribe` is the downstream subscriber, and each operator creates its own instance of subscriber that links the upstream (`source`) and downstream (`actual`).

**Comment by madgnome** [comment]
🤦‍♂️ Makes sense, thanks for the explanation!

**Comment by omayevskiy** [comment]
A proper solution would to be **NOT** to use MDC. Reactor and ThreadLocal (MDC is based on ThreadLocals) concepts does not work well together. 

Reactor comes with an Context feature that replaces the ThreadLocal but applied to a Flux or a Mono.

```
String key = "message";
Mono<String> r = Mono.just("Hello")
                .flatMap( s -> Mono.subscriberContext()
                                   .map( ctx -> s + " " + ctx.get(key)))
                .subscriberContext(ctx -> ctx.put(key, "World"));
```


the problem now is that we don't have a reactive logging, which would be aware of Mono, Flux and reactor's subscriber context. 

For instance the ```org.springframework.web.reactive.function.client.WebClient``` works native with Mono  and Flux, thus we can intercept the Context from the Mono to underlying http calls. 

```
WebClient c = WebClient.builder()
    .filter((request, next) -> Mono.subscriberContext()
            .flatMap(ctx -> {
                        ClientRequest cr = ClientRequest.from(request)
                                // manipulate request using context
                                .build();
                        return next.exchange(cr);
                    }
                ))
        .build();

webTestClient.post().uri("/api/repos")                
                .exchange() <-- a Mono
```

In the similar way there should be ideally a Reactive Logger giving us access to reactors Mono.subscriberContext() instead of ThreadLocals.

**Comment by vikasra77** [comment]
@ahus1 , @simonbasle Good conversation. Really liked the sample provided in your post above. Have one question. When I try to implement logging by putting in key-value pairs in the context, the logs get generated in reverse order due to the signal propagation. I did not find any way to have the timestamp in the order of execution, like 'receive event', 'call DB', 'generate response', 'send response' and so on. The first thing I see when I log to console is "send response with the first timestamp. Can you please help me here? Thanks

**Comment by marcingrzejszczak** [comment]
Can someone tell me why Sleuth isn't an answer to this problem?

**Comment by nanachimi** [comment]
I was not able to customized Sleuth to fit my goal. We have an existing logging format and this require  to a `correlation_id`. All `X-`Fields must be removed from the logs.

Maybe having more simple example base on concrete use-cases may help to onboarding

**Comment by tran302** [comment]
Is there any plan on making WebClient MDC-aware? Having to pass Context and put MDC for each log in every WebClient call does not seem like a feasible approach (and it does not account for unexpected error logs occurred from elsewhere in the call chain).

**Comment by michael-barker** [comment]
@tran302 this should already work for WebClient. Check out the whitelisted-mdc-key property.
https://cloud.spring.io/spring-cloud-sleuth/single/spring-cloud-sleuth.html#prefixed-fields

**Comment by tran302** [comment]
thanks.

**Comment by arunim29** [comment]
@simonbasle The context works on a publisher chain. If I need contextual logging in http api, I can do it using Context because a new publisher chain is made on every call. But if I am using web socket, the publisher chain is made only once. How to change context on every packet received from the client?

**Comment by sumitahuja17** [comment]
After googling on this topic for a while , I came across this thread. Do we have a sleuth based alternate solution to https://simonbasle.github.io/2018/02/contextual-logging-with-reactor-context-and-mdc/?
